library/core/src/ptr/mod.rs

```diff
@@ -1548,20 +1548,46 @@ pub unsafe fn write_volatile<T>(dst: *mut T, src: T) {
 
 /// Align pointer `p`.
 ///
-/// Calculate offset (in terms of elements of `stride` stride) that has to be applied
+/// Calculate offset (in terms of elements of `size_of::<T>()` stride) that has to be applied
 /// to pointer `p` so that pointer `p` would get aligned to `a`.
 ///
+/// # Safety
+/// `a` must be a power of two.
+#[lang = "align_offset"]
+#[rustc_do_not_const_check]
+#[cfg(not(bootstrap))]
+pub(crate) const unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usize {
+    // SAFETY: Caller ensures that `a` is a power of two.
+    unsafe { const_align_offset::<T>(p.addr(), a) }
+}
+
+#[lang = "align_offset"]
+#[cfg(bootstrap)]
+pub(crate) unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usize {
+    // SAFETY: Caller ensures that `a` is a power of two.
+    unsafe { const_align_offset::<T>(p.addr(), a) }
+}
+
+/// Align address `addr`.
+///
+/// Calculate offset (in terms of elements of `size_of::<T>()` stride) that has to be applied
+/// to address `addr` so that `addr` would get aligned to `a`.
+///
 /// Note: This implementation has been carefully tailored to not panic. It is UB for this to panic.
 /// The only real change that can be made here is change of `INV_TABLE_MOD_16` and associated
 /// constants.
 ///
+/// # Safety
+/// `a` must be a power of two.
+///
 /// If we ever decide to make it possible to call the intrinsic with `a` that is not a
 /// power-of-two, it will probably be more prudent to just change to a naive implementation rather
 /// than trying to adapt this to accommodate that change.
 ///
 /// Any questions go to @nagisa.
-#[lang = "align_offset"]
-pub(crate) unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usize {
+#[cfg_attr(not(bootstrap), lang = "const_align_offset")]
+#[rustc_allow_const_fn_unstable(const_exact_div)]
+pub(crate) const unsafe fn const_align_offset<T: Sized>(addr: usize, a: usize) -> usize {
```

Comment:
Why does this even need a separate const version of the function? I don't think a magic hook is needed.

Reply:
We need the magic hook to convert the pointer to an "address" to compute the modular multiplicative inverse (align offset) on. Alternatively we could just have one `const fn align_offset(p: *const T, a: usize)` and either 
* change the pointer that may have an address to a pointer that definitely has an "address" (offset in its allocation) in our hook and then `transmute` the address out of the pointer in the `align_offset` impl (which is documented UB); or
* put the entire modular multiplicative inverse computation in const eval. Even the [trivial implementation](https://github.com/rust-lang/rust/blob/51474dcba70a4391afff580bee86cb020022452d/library/core/tests/ptr.rs#L424-L430) would have to emit a lot of operations then instead of a single call.

Reply:
Sorry, what I meant is: yes we need to "intercept" these calls, but I think after doing the processing we can just pass through to the original implementation. We don't need a separate `const_align_offset`.

That is what Miri used to do (and still does with `-Zmiri-symbolic-alignment-check`).

Reply:
I've removed `const_align_offset` now, but I still hate [this `transmute`](https://github.com/rust-lang/rust/pull/102795/files#diff-c175d4e27676febf62c061d31cf9756d256b46e2e44cc6b3177d4ff75e932567R1628-R1631), especially since the error for `p as usize` says "avoiding this restriction via `transmute`, `union`, or raw pointers leads to compile-time undefined behavior".

Maybe we should adjust the documented behavior of const transmute from pointer to integers to be allowed on [invalid pointers](https://doc.rust-lang.org/nightly/core/ptr/fn.invalid.html)?

Reply:
Yeah transmute on pointers that are created from integers is fine, we should probably update the docs.

That's still a *huge* diff, why is so much code being moved/changed? I think it should work with a basically unchanged standard library, except for that transmute.

Reply:
A bunch of functions are duplicated now because they are only const if `not(boostrap)`. Is there any better way to make a function `const` based on a `cfg`?